keras_cv/ops/sampling.py

```diff
@@ -0,0 +1,85 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+
+def balanced_sample(
+    positive_matches: tf.Tensor,
+    negative_matches: tf.Tensor,
+    num_samples: int,
+    positive_fraction: float,
+):
+    """
+    Sampling ops to balance positive and negative samples, deals with both
+    batched and unbatched inputs.
+
+    Args:
+      positive_matches: [N] or [batch_size, N] boolean Tensor, True for
+        indicating the index is a positive sample
+      negative_matches: [N] or [batch_size, N] boolean Tensor, True for
+        indicating the index is a negative sample
+      num_samples: int, representing the number of samples to collect
+      positive_fraction: float. 0.5 means positive samples should be half
+        of all collected samples.
+
+    Returns:
+      selected_indicators: [`num_samples`] or [batch_size, `num_samples`]
+        integer Tensor, 1 for indicating the index is sampled, 0 for
+        indicating the index is not sampled.
+    """
+
+    def balanced_single_sample(positive_match, negative_match):
```

Comment:
Readability wise, maybe its better to lift this one out and have it as `_balanced_single_sample` in this file.

Reply:
it uses some other args within the function, such as `positive_fraction`, which cannot be unbatched during tf.vectorized_map